{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Config. Variables\n",
      "  Controlled Data: 100 %\n",
      "  Percentage of Bact. data: 30.0 %\n",
      "  Corruption Rate: 0.0 %\n",
      "  Ditance Limit: 0\n",
      "  Db string min lenght: 10\n",
      "  Db string max lenght: 200\n",
      "  Db max size: 15\n",
      "  Lake string min lenght: 10\n",
      "  Lake string max lenght: 100\n",
      "  Lake data max size: 10 \n",
      "\n",
      "\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '../database/lake_data.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-1952bb9b3c69>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m    240\u001b[0m \u001b[0mrandom\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mseed\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    241\u001b[0m \u001b[1;31m#size = random.randint(2, lakeSize)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 242\u001b[1;33m \u001b[0msize\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msum\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;36m1\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mopen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m'../database/lake_data.txt'\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    243\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    244\u001b[0m \u001b[1;31m#generates the database representation\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '../database/lake_data.txt'"
     ]
    }
   ],
   "source": [
    "from random import choice\n",
    "from string import ascii_uppercase\n",
    "import math\n",
    "import time\n",
    "from swalign import swalign\n",
    "from scipy.stats import beta\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "from scipy.integrate import simps\n",
    "from numpy import trapz\n",
    "import random\n",
    "import operator\n",
    "\n",
    "\n",
    "# ****** functions ********\n",
    "\n",
    "#creates dictionary with all permutations of length n \n",
    "#with repetition to index Feature Vector\n",
    "def CreateDictionary(n):\n",
    "    chars = \"ACGT\"\n",
    "    arr = list(itertools.product(chars, repeat=n))\n",
    "    \n",
    "    D = {}\n",
    "    i = 0\n",
    "\n",
    "    for a in arr:\n",
    "        D[''.join(a)] = i\n",
    "        i += 1\n",
    "        \n",
    "    return D\n",
    "\n",
    "#builds the feature vector for sequence using specified indexing dictionary\n",
    "def FeatureVector(dictionary, sequence, n):    \n",
    "    sLen = len(sequence)\n",
    "    arr = [0]*4**n\n",
    "    i = 0\n",
    "    \n",
    "    while(1):\n",
    "        w = sequence[i:i+n]\n",
    "        arr[D[w]] += 1\n",
    "        i += 1\n",
    "        if(i+n > sLen):\n",
    "            break\n",
    "    \n",
    "    return arr\n",
    "\n",
    "#Reads the DB files and puts the information of the file in a array of strings\n",
    "def readfile(filename):\n",
    "    temp = open(filename, 'r').read().split('\\n')\n",
    "    return temp\n",
    "    \n",
    "    \n",
    "#returns a random string of specified length\n",
    "#length: strign length\n",
    "def randomword(length):\n",
    "    return (''.join(choice('ACGT') for i in range(0, length)))\n",
    "\n",
    "#retuns an array of random strings\n",
    "#size: how many strings there will be in the array\n",
    "#lakeMinLen: min sequence length\n",
    "#lakeMaxLen: max sequence length\n",
    "def lakeString(size, lakeMinLen, lakeMaxLen):     \n",
    "    lake_water = []\n",
    "    for i in range(0, size):\n",
    "        random.seed()\n",
    "        #generates a random sequence length\n",
    "        y = random.randint(lakeMinLen, lakeMaxLen)\n",
    "        \n",
    "        _str = randomword(y)\n",
    "        lake_water.append(_str)\n",
    "    return lake_water\n",
    "\n",
    "#returns one array of strings that represents database data set\n",
    "#dbMinSeqLen: min string length\n",
    "#dbMaxSeqLen: max string length\n",
    "#dbSize: max number sequences in the array\n",
    "def dbStrings(dbMinSeqLen, dbMaxSeqLen, dbSize):\n",
    "    _array = []\n",
    "    #generates how many different sequences there will be for the array\n",
    "    random.seed()\n",
    "    x = random.randint(2, dbSize)\n",
    "\n",
    "    for i in range(0, x):\n",
    "        #generates a random sequence size\n",
    "        random.seed()\n",
    "        y = random.randint(dbMinSeqLen, dbMaxSeqLen)\n",
    "        _str = randomword(y)\n",
    "        _array.append(_str)\n",
    "    return _array\n",
    "\n",
    "#corrupts the string if it is under the corruption rate\n",
    "def stringCorruption(strg, corrupt_rate):\n",
    "    _str = ''\n",
    "    for k in strg:\n",
    "        #get a random number between 0 and 1\n",
    "        random.seed()\n",
    "        floatr = random.random()\n",
    "        if  floatr > corrupt_rate :\n",
    "            _str = _str + k\n",
    "        else:\n",
    "            _str = _str + choice('ACGT')     \n",
    "    return _str\n",
    "\n",
    "#configures the controlled data\n",
    "#parameters:\n",
    "#k_v: known_virus\n",
    "#k_b: known_bacterias\n",
    "#nControlled: how many records will be controlled\n",
    "#error: corruption limit\n",
    "def dataController(k_v, k_b, error, nControlVir, nControlBact, lakeMinLen, lakeMaxLen):\n",
    "    controlled_lake = []\n",
    "    vLen = len(known_viruses)\n",
    "    bLen = len(known_bacterias)\n",
    "    \n",
    "    #creates the specified amount of controlled lake sequences with viral pieces\n",
    "    for j in range (0, nControlVir):\n",
    "        #get a random virus\n",
    "        vir = known_viruses[random.randint(1, vLen-1)]\n",
    "        #get the length of the random virus\n",
    "        virLen = len(vir)            \n",
    "        count = 0\n",
    "        #tries to find valid random substrings to be used\n",
    "        while(True):\n",
    "            num1 = random.randint(0, virLen)\n",
    "            num2 = num1 + random.randint(lakeMinLen, lakeMaxLen)\n",
    "            if num2 > virLen-1:\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "        if num1 < num2:\n",
    "            #controls the max length\n",
    "            if(num2 - num1) > lakeMaxLen:\n",
    "                num2 = num1 + lakeMaxLen\n",
    "                \n",
    "            #get a substring\n",
    "            strg = vir[num1:num2]\n",
    "            #corrupts the string if it's under the error rate\n",
    "            strg = stringCorruption(strg, error)\n",
    "            controlled_lake.append(strg)\n",
    "        elif num2 < num1:\n",
    "            #controls the max length\n",
    "            if(num1 - num2) > lakeMaxLen:\n",
    "                num1 = num2 + lakeMaxLen\n",
    "                \n",
    "            #get a substring\n",
    "            strg = vir[num2:num1]\n",
    "            #corrupts the string if it's under the error rate\n",
    "            strg = stringCorruption(strg, error)\n",
    "            controlled_lake.append(strg)\n",
    "        else:\n",
    "            #controls the max length\n",
    "            if(virLen - num2) > lakeMaxLen:\n",
    "                num2 += lakeMaxLen\n",
    "            else:\n",
    "                num2 = virLen-1\n",
    "                \n",
    "            #get a substring\n",
    "            strg = vir[num1:num2]\n",
    "            #corrupts the string if it's under the error rate\n",
    "            strg = stringCorruption(strg, error)\n",
    "            controlled_lake.append(strg)  \n",
    "    \n",
    "    #creates the specified amount of controlled lake sequences with bacterial pieces\n",
    "    for j in range (0, nControlBact):\n",
    "        bact = known_bacterias[random.randint(1, bLen-1)]\n",
    "        bactLen = len(bact)            \n",
    "        count = 0\n",
    "        while(True):\n",
    "            num1 = random.randint(0, bactLen)\n",
    "            num2 = num1 + random.randint(lakeMinLen, lakeMaxLen)\n",
    "            if num2 > bactLen-1:\n",
    "                continue\n",
    "            else:\n",
    "                break\n",
    "        if num1 < num2:\n",
    "            #controls the max length\n",
    "            if(num2 - num1) > lakeMaxLen:\n",
    "                num2 = num1 + lakeMaxLen\n",
    "                \n",
    "            #get a substring\n",
    "            strg = bact[num1:num2]\n",
    "            #corrupts the string if it's under the error rate\n",
    "            strg = stringCorruption(strg, error)\n",
    "            controlled_lake.append(strg)\n",
    "        elif num2 < num1:\n",
    "            #controls the max length\n",
    "            if(num1 - num2) > lakeMaxLen:\n",
    "                num1 = num2 + lakeMaxLen\n",
    "                \n",
    "            #get a substring\n",
    "            strg = bact[num2:num1]\n",
    "            #corrupts the string if it's under the error rate\n",
    "            strg = stringCorruption(strg, error)\n",
    "            controlled_lake.append(strg)\n",
    "        else:\n",
    "            #controls the max length\n",
    "            if(virLen - num2) > lakeMaxLen:\n",
    "                num2 += lakeMaxLen\n",
    "            else:\n",
    "                num2 = bactLen-1\n",
    "                \n",
    "            #get a substring\n",
    "            strg = bact[num1:num2]\n",
    "            #corrupts the string if it's under the error rate\n",
    "            strg = stringCorruption(strg, error)\n",
    "            controlled_lake.append(strg)\n",
    "    \n",
    "    return controlled_lake\n",
    "    \n",
    "# ******************************************************* main ****************************************************\n",
    "\n",
    "####---------------- config. variables --------------------------------####\n",
    "percentage = 1      #how many will have virus/bacteria pieces\n",
    "bacterialRate = 0.3 #how many of the controlled lake sequences will bacterial piece\n",
    "error = 0.0         #corruption rate\n",
    "distLimit = 0       #distance between matching sequences\n",
    "\n",
    "lakeSize = 10    #max number of lake sequences\n",
    "lakeMinLen = 10     #min lake string length\n",
    "lakeMaxLen = 100     #max lake string length\n",
    "\n",
    "dbSize = 15  #max number sequences in the array\n",
    "dbMaxSeqLen = 200  #max db string length\n",
    "dbMinSeqLen = lakeMinLen #min db string length\n",
    "\n",
    "#print configuration\n",
    "print(\"\\n\\nConfig. Variables\")\n",
    "print(\"  Controlled Data:\", percentage*100, \"%\")\n",
    "print(\"  Percentage of Bact. data:\", bacterialRate*100, \"%\")\n",
    "print(\"  Corruption Rate:\", error*100, \"%\")\n",
    "print(\"  Ditance Limit:\", distLimit)\n",
    "print(\"  Db string min lenght:\", dbMinSeqLen)\n",
    "print(\"  Db string max lenght:\", dbMaxSeqLen)\n",
    "print(\"  Db max size:\", dbSize)\n",
    "print(\"  Lake string min lenght:\", lakeMinLen)\n",
    "print(\"  Lake string max lenght:\", lakeMaxLen)\n",
    "print(\"  Lake data max size:\", lakeSize, \"\\n\\n\")\n",
    "\n",
    "#generates how many different sequences there will be for the lake\n",
    "random.seed()\n",
    "#size = random.randint(2, lakeSize)\n",
    "size = sum(1 for line in open('../database/lake_data.txt'))\n",
    "\n",
    "#generates the database representation\n",
    "#known_viruses = dbStrings(dbMinSeqLen, dbMaxSeqLen, dbSize)\n",
    "#known_bacterias = dbStrings(dbMinSeqLen, dbMaxSeqLen, dbSize)\n",
    "known_viruses = readfile(\"../database/virus_db.txt\")\n",
    "known_bacterias = readfile(\"../database/bact_db.txt\")\n",
    "\n",
    "#generates the lake sequences w/ the specified percentage of controlled data\n",
    "P = math.ceil(size * percentage)\n",
    "lake = lakeString(size - P, lakeMinLen, lakeMaxLen) #produce a part with totally random sequences  \n",
    "\n",
    "#calculates how many sequences will be viral and how many will be bacterial\n",
    "nControlBact = round(P * bacterialRate)\n",
    "if bacterialRate == 0.0:\n",
    "   nControlVir = P\n",
    "else:\n",
    "   nControlVir = round((nControlBact * (1 - bacterialRate)) / bacterialRate)\n",
    "\n",
    "print(\"# inserted viral pieces:\", nControlVir, \"\\n# inserted bacterial pieces: \", nControlBact)\n",
    "size = size - P + nControlVir + nControlBact\n",
    "print(\"True bacterial percentage:\", nControlBact/size*100, \"%\")\n",
    "\n",
    "#produce the other part with viral or bacterial pieces\n",
    "controlled = dataController(known_viruses, known_bacterias, error, nControlVir, nControlBact, lakeMinLen, lakeMaxLen)\n",
    "#lake = lake + controlled #join the sequences\n",
    "lake = readfile(\"../database/lake_data.txt\")\n",
    "\n",
    "#print sequences\n",
    "#i = 0\n",
    "#print(\"*** Lake Sequences\")\n",
    "#for w in lake:\n",
    "#    print(i, \" \", w)\n",
    "#    i = i + 1 \n",
    "#    \n",
    "#print(\"\\n\\n*** Viral Sequences\")\n",
    "#m = 0\n",
    "#for w in known_viruses:\n",
    "#    print(m, \" \", w)\n",
    "#    m = m + 1 \n",
    "#    \n",
    "#print(\"\\n\\n*** Baterial Sequences\")\n",
    "#m = 0\n",
    "#for w in known_bacterias:\n",
    "#    print(m, \" \", w)\n",
    "#    m = m + 1 \n",
    "#print(\"\\n\\n\")\n",
    "\n",
    "#matrix with all feature vectors\n",
    "n = 4\n",
    "D = CreateDictionary(n)\n",
    "matrix = []\n",
    "\n",
    "for w in lake:\n",
    "    arr = FeatureVector(D, w, n)\n",
    "    matrix.append(arr)\n",
    "   \n",
    "for w in known_viruses:\n",
    "    arr = FeatureVector(D, w, n)\n",
    "    matrix.append(arr)\n",
    "   \n",
    "for w in known_bacterias:\n",
    "    arr = FeatureVector(D, w, n)\n",
    "    matrix.append(arr)\n",
    "\n",
    "\n",
    "nBact = 0    #final counter\n",
    "nVirus = 0   #final counter\n",
    "nUnknown = 0 #final counter\n",
    "_y = 1 #final curve\n",
    "\n",
    "#Local Alignment settings\n",
    "# choose your own values hereâ€¦ 2 and -1 are common.\n",
    "match = 1                                                  #scores 1 point for matching letters\n",
    "mismatch = -1                                              #looses 1 point for mismatching letters\n",
    "scoring = swalign.NucleotideScoringMatrix(match, mismatch)\n",
    "sw = swalign.LocalAlignment(scoring)                       # you can also choose gap penalties, etc...\n",
    "\n",
    "#draws beta distributions\n",
    "x = np.arange(0, 1, 0.001)\n",
    "\n",
    "start = time.clock() #start clock to measure proccessing time\n",
    "\n",
    "for l in lake:       #for each sequece in the lake\n",
    "    #print(\"***l len: \", len(l))\n",
    "    viralMatches = 0 \n",
    "    bactMatches = 0\n",
    "    lLen = len(l)                                  #get sequence length\n",
    "    #gives different weights depending of the sequence length \n",
    "    pnts = math.ceil((lLen/dbMaxSeqLen*100) + 1)\n",
    "    _limit = round(lLen * (100 - distLimit) / 100) #get the minimum points to be considered a match\n",
    "    for v in known_viruses:                        #compare the sequence with all virus in the database\n",
    "        alignment = sw.align(v, l)\n",
    "        if alignment.score >= _limit:\n",
    "            viralMatches += pnts       \n",
    "    for b in known_bacterias: #compare the sequence with all bacterias in the database\n",
    "        alignment = sw.align(b, l)\n",
    "        if alignment.score >= _limit:\n",
    "            bactMatches += pnts\n",
    "\n",
    "    #draws beta distributions    \n",
    "    bactMatches += 1\n",
    "    viralMatches += 1    \n",
    "    mean, var, skew, kurt = beta.stats(bactMatches, viralMatches, moments='mvsk')\n",
    "    #_label = \"Viral:\", viralMatches, \"Bacterial:\", bactMatches\n",
    "    #ax.plot(x, beta.pdf(x, bactMatches, viralMatches),'p-', lw=3, alpha=0.6, label=_label)\n",
    "    y = beta.pdf(x, bactMatches, viralMatches)\n",
    "    yIndex = np.argmax(_y)\n",
    "\n",
    "    #multiplies the coordinates\n",
    "    _y *= y \n",
    "\n",
    "    \n",
    "end = time.clock()\n",
    "print(\"\\n\\n>>>Elapsed Time:\", round(end - start, 3))    \n",
    "\n",
    "#Gets the highest value, the peak value\n",
    "yIndex = np.argmax(_y)\n",
    "\n",
    "print(\"\\n\\nReport\")\n",
    "print(\"   Viruses:\", round((1 - x[yIndex]) * 100, 2), \"%\")\n",
    "print(\"   Bacteria:\", round(x[yIndex] * 100, 2), \"%\")\n",
    "\n",
    "#plots the multiplication between the beta distribution curves \n",
    "fig, ax2 = plt.subplots(1, 1)\n",
    "ax2.plot(x, _y, 'y-', lw = 5, alpha=0.9, label=\"label\")\n",
    "ax2.set_title(\"Beta Distribution Multiplication\")\n",
    "ax2.set_xlabel(\"Bacterial Rate\")\n",
    "ax2.legend(loc='best', frameon=False)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
